tests/mve/gtest_scene.cc

```diff
@@ -0,0 +1,751 @@
+    EXPECT_FALSE(scene_with_dirty_bundle_and_clean_views->is_dirty());
+}
+
+TEST(SceneTest,
+     save_bundle_does_not_clean_a_scene_if_any_of_its_views_is_dirty)
+{
+    OnScopeExit clean_up;
+
+    mve::Scene::Ptr scene_with_dirty_view = [&clean_up]() {
+        std::string path_to_scene =
+                create_scene_on_disk(7, make_bundle(8), clean_up);
+        mve::Scene::Ptr scene = mve::Scene::create(path_to_scene);
+        make_a_clean_view_dirty(scene);
+        scene->set_bundle(mve::Bundle::create());
+        return scene;
+    }();
+    scene_with_dirty_view->save_views();
+    EXPECT_TRUE(scene_with_dirty_view->is_dirty());
+}
+
+TEST(SceneTest,
+     reset_bundle_cleans_a_scene_if_only_its_bundle_is_dirty)
+{
+    OnScopeExit clean_up;
+
+    mve::Scene::Ptr scene_with_dirty_bundle_and_clean_views = [&clean_up]() {
+        std::string path_to_scene =
+                create_scene_on_disk(10, make_bundle(3), clean_up);
+        mve::Scene::Ptr scene = mve::Scene::create(path_to_scene);
+        scene->set_bundle(mve::Bundle::create());
+        return scene;
+    }();
+    scene_with_dirty_bundle_and_clean_views->reset_bundle();
+    EXPECT_FALSE(scene_with_dirty_bundle_and_clean_views->is_dirty());
+}
+
+TEST(SceneTest,
+     reset_bundle_does_not_clean_a_scene_if_any_of_its_views_is_dirty)
+{
+    OnScopeExit clean_up;
+
+    mve::Scene::Ptr scene_with_dirty_bundle_and_clean_views = [&clean_up]() {
+        std::string path_to_scene =
+                create_scene_on_disk(10, make_bundle(3), clean_up);
+        mve::Scene::Ptr scene = mve::Scene::create(path_to_scene);
+        make_a_clean_view_dirty(scene);
+        scene->set_bundle(mve::Bundle::create());
+        return scene;
+    }();
+    scene_with_dirty_bundle_and_clean_views->reset_bundle();
+    EXPECT_TRUE(scene_with_dirty_bundle_and_clean_views->is_dirty());
+}
+
+TEST(SceneTest,
+     saving_the_dirty_views_of_a_scene_cleans_the_scene_if_its_bundle_is_clean)
+{
+    using View = mve::View::Ptr;
+    OnScopeExit clean_up;
+
+    mve::Scene::Ptr scene_with_dirty_views_and_clean_bundle = [&clean_up]() {
+        std::string path_to_scene =
+                create_scene_on_disk(10, make_bundle(6), clean_up);
+        mve::Scene::Ptr scene = mve::Scene::create(path_to_scene);
+        for(int i = 0; i < 5; ++i)
+            make_a_clean_view_dirty(scene);
+        return scene;
+    }();
+
+    for (View& view : scene_with_dirty_views_and_clean_bundle->get_views())
+        if (view->is_dirty())
+            view->save_view();
+
+    EXPECT_FALSE(scene_with_dirty_views_and_clean_bundle->is_dirty());
+}
+
+//== End of tests ==============================================================
+
+namespace {
+
+std::string
+create_scene_on_disk(const std::size_t view_count,
+                     const mve::Bundle::Ptr bundle,
+                     OnScopeExit& on_scope_exit)
+{
+    namespace fs = util::fs;
+
+    std::string scene_directory { std::tmpnam(nullptr) };
```

Comment:
Don't ever use {} for assignment. The {} operator is pretty evil. Consider this, what do you think this is doing?

    int a = 1, b = 10, c = 20;
    std::vector<int> test {&a, &c};